src/tree.rs

```diff
@@ -241,24 +264,126 @@ where
-        match n.ntype {
-            NodeType::Leaf(_) => {
-                self.elements.remove(element);
-                self.map.remove(element);
-                n.ntype = NodeType::Empty;
-                if let Some((element, node)) = self.nodes.collapse(parent)? {
-                    self.map[element] = node;
+    pub fn remove(&mut self, elem: ElementId) -> Result<(), TreeError> {
+        if let Some(element) = self.get_element(elem) {
+            let volume = element.volume();
+            if self.nodes[self.root].aabb.overlaps(&volume) {
+                let mut removals: SmallVec<[Removal; 16]> = SmallVec::new();
+                removals.push(Removal {
+                    parent: None,
+                    node: self.root,
+                });
+                while let Some(removal) = removals.pop() {
+                    self._remove(elem, volume, removal, &mut removals)?;
+                }
+                self.elements.tombstone(elem);
+                Ok(())
+            } else {
+                return Err(TreeError::OutOfTreeBounds(format!(
+                    "{volume} is outside of aabb: min: {} max: {}",
+                    self.nodes[self.root].aabb.min, self.nodes[self.root].aabb.max,
+                )));
+            }
+        } else {
+            Err(TreeError::ElementNotFound(format!(
+                "Element with id: {} not found",
+                elem.0
+            )))
+        }
+    }
+
+    #[inline]
+    fn _remove(
+        &mut self,
+        element: ElementId,
+        volume: Aabb<U>,
+        removal: Removal,
+        removals: &mut SmallVec<[Removal; 16]>,
+    ) -> Result<(), TreeError> {
+        let Removal { parent, node } = removal;
+
+        if self.nodes.is_garbaged(node) {
+            return Ok(());
+        }
+
+        let ntype = self.nodes[node].ntype;
+        match ntype {
+            NodeType::Empty => Ok(()),
+
+            NodeType::Leaf(e) if e == element => {
+                self.nodes[node].ntype = NodeType::Empty;
+                if let Some(parent) = parent {
+                    self.nodes.maybe_collapse(parent);
+                }
+                Ok(())
+            }
+
+            NodeType::Leaf(_) => Ok(()),
+
+            NodeType::Branch(branch) => {
+                let branch_center = branch.center(&self.nodes);
+                if volume.min.x <= branch_center.x {
+                    if volume.min.y <= branch_center.y {
+                        if volume.min.z <= branch_center.z {
+                            removals.push(Removal {
+                                parent: Some(node),
+                                node: branch.x0_y0_z0(),
+                            });
+                        }
+                        if volume.max.z >= branch_center.z {
+                            removals.push(Removal {
+                                parent: Some(node),
+                                node: branch.x0_y0_z1(),
```

Comment:
Can we put that code repeating in `fn _insert` in a separate function and mark it `inline`?

Reply:
Done